core/storage/wal.rs

```diff
@@ -1012,6 +1129,139 @@ impl WalFile {
     fn get_shared(&self) -> &mut WalFileShared {
         unsafe { self.shared.get().as_mut().unwrap() }
     }
+
+    // Coordinate what the maximum safe frame is for us to backfill when checkpointing.
```

Comment:
Add:

```
/// Coordinate what the maximum safe frame is for us to backfill when checkpointing.
/// We can never backfill a frame with a higher number than any reader's max frame,
/// because we might overwrite content the reader is reading from the database file.